core/state/trie_prefetcher.go

```diff
@@ -188,20 +250,25 @@ func (p *triePrefetcher) trie(root common.Hash) Trie {
 	if p.fetches != nil {
 		trie := p.fetches[root]
 		if trie == nil {
-			p.deliveryMissMeter.Mark(1)
 			return nil
 		}
 		return p.db.CopyTrie(trie)
 	}
-	// Otherwise the prefetcher is active, bail if no trie was prefetched for this root
+
+	p.fetchersMutex.RLock()
 	fetcher := p.fetchers[root]
+	p.fetchersMutex.RUnlock()
 	if fetcher == nil {
 		p.deliveryMissMeter.Mark(1)
 		return nil
 	}
+
 	// Interrupt the prefetcher if it's by any chance still running and return
 	// a copy of any pre-loaded trie.
-	p.abortChan <- fetcher // safe to do multiple times
+	select {
+	case <-p.closeAbortChan:
+	case p.abortChan <- fetcher: // safe to do multiple times
```

Comment:
I agree with @qinglin89 , if we continue to send message to `p.abortChan` when `abortLoop` already return , deadlock may happen.


I am afraid about this, a long run test  and a go routine profile to ensure no routine leakage happen.

Reply:
As mention above, should no deadlock. I can capture a go profile, I have test device running for more than 1day.

Reply:
get.